[C++20] Destroying delete and deleted destructors

AaronBallman opened the pull request:
When a destroying delete overload is selected, the destructor is not automatically called. Therefore, the destructor can be deleted without causing the program to be ill-formed.

Fixes #46818

AaronBallman committed:
[C++20] Destroying delete and deleted destructors

When a destroying delete overload is selected, the destructor is not
automatically called. Therefore, the destructor can be deleted without
causing the program to be ill-formed.

cor3ntin commented on clang/test/CXX/expr/expr.unary/expr.delete/p10.cpp:
```diff
@@ -23,3 +30,14 @@ struct alignas(Align) B {
+
+// Ensure that a deleted destructor is acceptable when the selected overload
+// for operator delete is a destroying delete. See the comments in GH118660.
+struct S {
+  ~S() = delete;
+  void operator delete(S *, std::destroying_delete_t) noexcept {}
+};
+
+void foo(S *s) {
+  delete s; // Was rejected, is intended to be accepted.
+}
```
can you add a test with a private destructor?
(I found this unresolved issue https://cplusplus.github.io/CWG/issues/2889.html that is tangentially relevant )

AaronBallman replied:
Good call, that identified an issue which I've now fixed.

erichkeane commented on clang/lib/Sema/SemaExprCXX.cpp:
```diff
@@ -3792,13 +3792,15 @@ Sema::ActOnCXXDelete(SourceLocation StartLoc, bool UseGlobal,
       }
 
-      if (!PointeeRD->hasIrrelevantDestructor())
+      if (!PointeeRD->hasIrrelevantDestructor() &&
```
Would love a standards quote here.

AaronBallman replied:
Done

erichkeane reviewed (commented):
1 nit, else LGTM.

AaronBallman committed:
Update based on review feedback

* Added a standards reference
* Added a test case
* Fixed an issue the new test case identified

AaronBallman committed:
Fix one of the two failing tests, update release notes

It turns out someone filed an issue for this.

AaronBallman committed:
Fix the other failing test case

A virtual destructor still needs to be accessible to a destroying
delete because the actual delete operator called depends on the dynamic
type of the pointer, not the static type.

zygoloid commented:
As I just noted in #118687, I think we also need similar treatment for `noexcept(delete p)`.

AaronBallman commented on clang/lib/Sema/SemaExprCXX.cpp:
```diff
@@ -3768,6 +3768,28 @@ Sema::ActOnCXXDelete(SourceLocation StartLoc, bool UseGlobal,
+    // expression will invoke the destructor (if any) for the object or the
+    // elements of the array being deleted.
+    //
+    // This means we should not look at the destructor for a destroying
+    // delete operator, as that destructor is never called, unless the
+    // destructor is virtual (see [expr.delete]p8.1) because then the
+    // selected operator depends on the dynamic type of the pointer.
```
Oof, good catch! Also, thank you for reporting this to Core.

Do you think I should make any changes here as part of this PR? I saw your comment on https://github.com/llvm/llvm-project/pull/118687#issuecomment-2521239072, I could handle that in a follow-up (probably tomorrow, so if you want me to revert those changes, I can). It's not quite clear to me whether this is a "wait for Core because we don't want behavioral churn" situation or not.

zygoloid replied:
I think for now it'd be reasonable to say that:

- if the destructor is virtual, `noexcept` looks only at the exception specification of the destructor
- otherwise, `noexcept` looks at the selected deallocation function and, if that function is not a destroying operator delete, also the destructor

That'd match what we do in lowering. I'm not sure how that lines up with other implementations, but it seems like the most important thing is that `noexcept` is consistent with whether an exception can actually escape from the evaluation.

AaronBallman committed:
Merge remote-tracking branch 'origin' into aballman-gh118660-second-bug

AaronBallman committed:
Update noexcept behavior based on review feedback

AaronBallman commented:
FYI, I'm going to be out on vacation after today, so if this is ready to land, someone else can feel free to press the button. Otherwise, I can pick this up again when I'm back.

AaronBallman commented:
Ping @zygoloid

zygoloid commented on clang/lib/Sema/SemaExceptionSpec.cpp:
```diff
@@ -1200,21 +1200,35 @@ CanThrowResult Sema::canThrow(const Stmt *S) {
+      // expression will invoke the destructor (if any) for the object or the
+      // elements of the array being deleted.
+      //
+      // So if the destructor is virtual, we only look at the exception
+      // specification of the destructor regardless of what operator delete is
+      // selected. Otherwise, we look at the selected operator delete, and if
```
This comment doesn't match the code, which always considers the `operator delete`.

Would it make sense to move `IsDtorCalled` onto `CXXDeleteExpr` so you can call it from here too?

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-gh118660-second-bug

AaronBallman committed:
Update based on review feedback

AaronBallman committed:
Backing out a whitespace change; NFC

zygoloid commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -2855,6 +2855,11 @@ class CXXDestructorDecl : public CXXMethodDecl {
+  /// Will this destructor ever be called when considering which deallocation
+  /// function is associated with the destructor? Can optionally be passed an
+  /// 'operator delete' function declaration to test against specifically.
+  bool isDestructorCalled(const FunctionDecl *OpDel = nullptr) const;
```
For readability of call sites, the name of this function should probably mention `delete` in some way. (`isCalledByDelete` maybe?)

AaronBallman replied:
Good call, I've renamed, thanks!

zygoloid reviewed (approved):
LG other than the function name.

AaronBallman committed:
Rename API for clarity

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-gh118660-second-bug
